File: ts/src/data/enemyDatabase.ts
/**
 * Normalize enemy combat databases from supported upstream archive shapes.
 *
 * The legacy archive wraps rows in ``{ enemies: [{ Key, Value }] }`` while
 * AKDP releases the equivalent ``{ enemyId: Value[] }`` mapping directly.
 */
import { parseLevel } from "./levelParser.js";

interface EnemyDatabaseValue<T> {
  level?: number | string;
  enemyData?: T;
}

type LegacyEnemyDatabase<T> = {
  enemies?: Array<{ Key?: string; Value?: EnemyDatabaseValue<T>[] }>;
};

function addRow<T>(
  index: Record<string, Record<number, T>>,
  enemyId: string,
  values: unknown,
): void {
  if (!Array.isArray(values)) return;
  const levelMap: Record<number, T> = {};
  for (const value of values) {
    if (typeof value !== "object" || value === null) continue;
    const row = value as EnemyDatabaseValue<T>;
    if (row.enemyData && typeof row.enemyData === "object") {
      levelMap[parseLevel(row.level)] = row.enemyData;
    }
  }
  if (Object.keys(levelMap).length > 0) index[enemyId] = levelMap;
}

/** Normalize a supported enemy_database.json payload into an ID/level index. */
export function normalizeEnemyDatabase<T>(raw: unknown): Record<string, Record<number, T>> {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    throw new Error("enemy_database.json 格式异常：根节点必须是对象。");
  }

  const root = raw as LegacyEnemyDatabase<T> & Record<string, unknown>;
  const index: Record<string, Record<number, T>> = {};
  if (Array.isArray(root.enemies)) {
    for (const row of root.enemies) {
      if (!row?.Key) continue;
      addRow(index, row.Key, row.Value);
    }
    return index;
  }

  for (const [enemyId, values] of Object.entries(root)) {
    addRow(index, enemyId, values);
  }
  if (Object.keys(index).length === 0) {
    throw new Error("enemy_database.json 格式异常：未找到敌人等级数据。");
  }
  return index;
}

/**
 * Project the raw level map onto the default (level-0) entry per enemy.
 *
 * Selection semantics replicate enemy.ts's historical projection verbatim
 * (nullish ``??``); the Python mirror uses truthy ``or`` instead — a known,
 * deliberate cross-implementation divergence, do not unify.
 */
export function level0Index<T>(
  levels: Record<string, Record<number, T>>,
): Record<string, T> {
  const index: Record<string, T> = {};
  for (const [enemyId, levelMap] of Object.entries(levels)) {
    const first = levelMap[0] ?? Object.values(levelMap)[0];
    if (first) index[enemyId] = first;
  }
  return index;
}

File: python/src/prts_mcp/data/enemy_database.py
"""Normalize supported upstream ``enemy_database.json`` archive shapes."""
from __future__ import annotations

from typing import Any


def normalize_enemy_database(raw: Any) -> dict[str, dict[int, dict[str, Any]]]:
    """Return an ``enemy ID -> level -> enemyData`` index for supported input."""
    if not isinstance(raw, dict):
        raise TypeError("enemy_database.json 格式异常：根节点必须是对象。")

    index: dict[str, dict[int, dict[str, Any]]] = {}

    def add_row(enemy_id: str, values: Any) -> None:
        if not isinstance(values, list):
            return
        level_map: dict[int, dict[str, Any]] = {}
        for value in values:
            if not isinstance(value, dict):
                continue
            try:
                level = int(value.get("level", 0))
            except (TypeError, ValueError):
                level = 0
            enemy_data = value.get("enemyData")
            if isinstance(enemy_data, dict):
                level_map[level] = enemy_data
        if level_map:
            index[enemy_id] = level_map

    legacy_rows = raw.get("enemies")
    if isinstance(legacy_rows, list):
        for row in legacy_rows:
            if not isinstance(row, dict):
                continue
            key = row.get("Key")
            if isinstance(key, str) and key:
                add_row(key, row.get("Value"))
        return index

    for enemy_id, values in raw.items():
        add_row(str(enemy_id), values)
    if not index:
        raise TypeError("enemy_database.json 格式异常：未找到敌人等级数据。")
    return index


def level0_index(levels: dict[str, dict[int, dict[str, Any]]]) -> dict[str, dict[str, Any]]:
    """Project the raw level map onto the default (level-0) entry per enemy.

    Selection semantics replicate enemy.py's historical projection verbatim
    (truthy-``or``: an empty-dict level-0 entry falls through to the first
    available level); the TS mirror uses nullish ``??`` instead — a known,
    deliberate cross-implementation divergence, do not unify.
    """
    return {
        enemy_id: level_map.get(0) or next(iter(level_map.values()))
        for enemy_id, level_map in levels.items()
        if level_map
    }
